geopandas/geodataframe.py

```diff
@@ -1634,7 +1634,17 @@ def _constructor_from_mgr(self, mgr, axes):
             return _geodataframe_constructor_with_fallback(
                 pd.DataFrame._from_mgr(mgr, axes)
             )
-        return GeoDataFrame._from_mgr(mgr, axes)
+        gdf = GeoDataFrame._from_mgr(mgr, axes)
+        # _from_mgr doesn't preserve metadata (expect __finalize__ to be called)
+        # still need to mimic __init__ behaviour with geometry=None
+        if (gdf.columns == "geometry").sum() == 1:  # only if "geometry" is single col
+            try:
+                gdf["geometry"] = _ensure_geometry(gdf["geometry"].values)
```

Comment:
Do we need this `_ensure_geometry` call? 
Yes, to preserve the exact behaviour of before (and so maybe best to do that for 0.x), but longer term, we could just set the `_geometry_column_name` (in this code path, we already checked that there is at least one column with geometry dtype, we just didn't check it is the one named "geometry")

Reply:
I think it's probably uncessary, but I don't know the pandas internals well enough to be sure - I would think if we get a blockmanager it's all or nothing, either all dtypes were conserved or they were all lost (I would expect e.g. apply doesn't come from _from_mgr, but I haven't checked)

Reply:
I've just tested that using 

```py
        if (gdf.columns == "geometry").sum() == 1:
            gdf._geometry_column_name = "geometry"
```
is enough to fix the example from the issue on pandas 2.1.4. So the try/except may not be needed here.

Reply:
Yeah it's definitely not needed for this example, it's moreso that by implementing `_constructor_from_mgr` there are pandas code paths in pandas which will use this instead of our `_constructor` which dispatches to `__init__`, and this behaviour is part of `__init__`. I will remove it for now then, there's a possibility this could lead to a weird error after some kind of groupby apply type operation, but we can deal with that if it comes up